### PR TITLE
Poll for machine deletion and kubeadm configmap updates

### LINKED ISSUE #120: Error removing node from kubeadm configmap
We aren't currently waiting for the machine to be deleted before we update the kubeadm configmap. This can lead to an error if the ELB directs traffic to the machine being deleted (for the kubeadm configmap update):
```
INFO[0072] Deleting etcd member                          id=3427fd0342505fdf
INFO[0072] Running etcdctl                               args="member remove 3427fd0342505fdf" pod=etcd-ip-10-0-0-59.ec2.internal
INFO[0072] etcdctl stdout: Member 3427fd0342505fdf removed from cluster 22b620bf07e549f6
INFO[0072] etcdctl stderr:
INFO[0072] Deleting existing machine                     name=cp3-0 namespace=default
Error: error getting kubeadm configmap from target cluster: Get https://test3-apiserver-2048175614.us-east-1.elb.amazonaws.com:6443/api/v1/namespaces/kube-system/configmaps/kubeadm-config: unexpected EOF
Get https://test3-apiserver-2048175614.us-east-1.elb.amazonaws.com:6443/api/v1/namespaces/kube-system/configmaps/kubeadm-config: unexpected EOF
error getting kubeadm configmap from target cluster
github.com/vmware/cluster-api-upgrade-tool/pkg/upgrade.(*ControlPlaneUpgrader).removeNodeFromKubeadmConfigMap
        /Users/andy/code/capi-dev/cluster-api-upgrade-tool/pkg/upgrade/control_plane.go:1006
github.com/vmware/cluster-api-upgrade-tool/pkg/upgrade.(*ControlPlaneUpgrader).updateMachine
        /Users/andy/code/capi-dev/cluster-api-upgrade-tool/pkg/upgrade/control_plane.go:489
github.com/vmware/cluster-api-upgrade-tool/pkg/upgrade.(*ControlPlaneUpgrader).updateMachines
        /Users/andy/code/capi-dev/cluster-api-upgrade-tool/pkg/upgrade/control_plane.go:582
github.com/vmware/cluster-api-upgrade-tool/pkg/upgrade.(*ControlPlaneUpgrader).Upgrade
        /Users/andy/code/capi-dev/cluster-api-upgrade-tool/pkg/upgrade/control_plane.go:204
main.main.func1
        /Users/andy/code/capi-dev/cluster-api-upgrade-tool/main.go:71
github.com/spf13/cobra.(*Command).execute
        /Users/andy/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826
github.com/spf13/cobra.(*Command).ExecuteC
        /Users/andy/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914
github.com/spf13/cobra.(*Command).Execute
        /Users/andy/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main
        /Users/andy/code/capi-dev/cluster-api-upgrade-tool/main.go:140
runtime.main
        /usr/local/Cellar/go/1.12.9/libexec/src/runtime/proc.go:200
runtime.goexit
        /usr/local/Cellar/go/1.12.9/libexec/src/runtime/asm_amd64.s:1337
 andy  ~  code  capi-dev  cluster-api-upgrade-tool   master … 5  ./bin/cluster-api-upgrade-tool --cluster-namespace default --cluster-name test3 --kubernetes-version 1.15.3 --upgrade-id 1
INFO[0000] Retrieving cluster from management cluster    cluster-name=test3 cluster-namespace=default
INFO[0000] Creating target kubernetes client
INFO[0000] Rerun with `--upgrade-id=1` if this upgrade fails midway and you want to retry
INFO[0000] Listing machines                              labelSelector="map[cluster.x-k8s.io/cluster-name:test3 cluster.x-k8s.io/control-plane:true]"
INFO[0000] Checking for existence of kubelet configmap   name=kubelet-config-1.15
INFO[0000] kubelet configmap already exists              name=kubelet-config-1.15
INFO[0000] Looking up role                               role-name="kube-system/kubeadm:kubelet-config-1.15"
INFO[0000] Looking up role binding                       role-name="kube-system/kubeadm:kubelet-config-1.15"
INFO[0000] Checking etcd health
INFO[0000] Running etcdctl                               args="member list -w json" pod=etcd-ip-10-0-0-59.ec2.internal
INFO[0000] etcdctl stdout: {"header":{"cluster_id":2501222647894198774,"member_id":8205522946952724276,"raft_term":8},"members":[{"ID":8205522946952724276,"name":"ip-10-0-0-59.ec2.internal","peerURLs":["https://10.0.0.59:2380"],"cli
entURLs":["https://10.0.0.59:2379"]}]}
INFO[0000] etcdctl stderr:
INFO[0000] Running etcdctl                               args="endpoint health --endpoints https://10.0.0.59:2379" pod=etcd-ip-10-0-0-59.ec2.internal
INFO[0000] etcdctl stdout: https://10.0.0.59:2379 is healthy: successfully committed proposal: took = 934.064µs
https://10.0.0.59:2379 is healthy: successfully committed proposal: took = 970.936µs
INFO[0000] etcdctl stderr:
INFO[0000] Updating provider IDs to nodes
INFO[0000] Updating provider IDs to nodes
INFO[0000] Updating kubernetes version
INFO[0000] Updating machines
INFO[0000] Running etcdctl                               args="member list -w json" pod=etcd-ip-10-0-0-59.ec2.internal
INFO[0001] etcdctl stdout: {"header":{"cluster_id":2501222647894198774,"member_id":8205522946952724276,"raft_term":8},"members":[{"ID":8205522946952724276,"name":"ip-10-0-0-59.ec2.internal","peerURLs":["https://10.0.0.59:2380"],"cli
entURLs":["https://10.0.0.59:2379"]}]}
INFO[0001] etcdctl stderr:
INFO[0001] Skipping machine as it is a replacement machine for the in-process upgrade  machine=default/cp3-0.upgrade.1 upgrade-id=1
INFO[0001] Removing upgrade annotations
```

We should do 2 things:

1) Wait up to _n_ minutes for the machine to be fully deleted
2) Retry the kubeadm configmap update up to 5 or so minutes

### LINKED ISSUE #121: Need to adjust handling timeouts during/after machine deletion
Our current logic is:

1. List all non-deleted control plane machines
1. For each machine
    1. Do all the work required to create a replacement Machine, etc.
    1. Remove original from etcd
    1. Delete original Machine
    1. Wait for deletion to finish
    1. Remove original node's entry from kubeadm configmap

If we get a timeout waiting for the deleted Machine to fully go away, or trying to update the kubeadm configmap, we won't ever reprocess the deleted Machine the next time the tool runs.

To correct this, we should:

1. List all machines
1. For each machine
    1. If it is active, perform the replacement logic (not including deletion)
    1. Regardless of status, perform deletion logic + configmap updating
        1. If deletion fails with a NotFound error, that's ok, proceed to configmap update

There is still a chance we could miss something in between upgrade runs. For example:

1. Try to delete machine - get a timeout
1. Wait some amount of time - machine gets actually deleted
1. Rerun tool - because machine is gone, we don't see it again, and and miss removing it from the kubeadm configmap

To mitigate, at the beginning and/or end of each invocation, we could update the kubeadm configmap so it matches the nodes that currently exist when we run `kubectl get nodes` against the workload cluster.

cc @chuckha @rudoi @williamsandrew for 👀 and feedback

### PR DESCRIPTION
**Pull request by ncdc**
Fixes #120

### CODE REVIEW

File: pkg/upgrade/control_plane.go [hunk 1]
```diff
@@ -482,12 +482,37 @@ func (u *ControlPlaneUpgrader) updateMachine(replacementKey ctrlclient.ObjectKey
 		return true, nil
 	})
 	if err != nil {
-		return errors.Wrapf(err, "timed out deleting machine %s/%s", machine.Namespace, machine.Name)
+		return errors.Wrapf(err, "timed out asking to delete machine %s/%s", machine.Namespace, machine.Name)
+	}
+
+	log.Info("Waiting for machine to be deleted")
+	err = wait.PollImmediate(deleteMachineInterval, deleteMachineTimeout, func() (bool, error) {
```
Comment (detiber) [inline pkg/upgrade/control_plane.go]: Would it make sense to use foreground deletion on the Machine instead of polling here?

Comment (ncdc) [reply pkg/upgrade/control_plane.go]: I think we still have to poll - doesn't the `DELETE` http request return immediately no matter what you use for propagation policy? The GC controller is responsible for dealing with foreground vs background deletion of dependents.

Comment (detiber) [reply pkg/upgrade/control_plane.go]: I want to say that the call blocks until it has been deleted, at least that is the behavior that I remember from implementing pivot for `clusterctl`

Comment (chuckha) [reply pkg/upgrade/control_plane.go]: I tried this out and the call did not block:

https://gist.github.com/chuckha/032086240945cfd3281b6137d7b4d1da

If this sample code waits for deletion to complete, there should be a panic.

Comment (ncdc) [reply pkg/upgrade/control_plane.go]: Yeah I'm 99.999% certain that you have to add code to block if you want to wait

**Pull request merged**